### PR TITLE
[CLIPPER-107][CLIPPER-109] Return JSON-formatted responses to prediction queries

### PR DESCRIPTION
**Pull request by Corey-Zumar**
Modifies query frontend and dependent infrastructure to return JSON-formatted responses to prediction queries.

From [CLIPPER-107](https://clipper.atlassian.net/projects/CLIPPER/issues/CLIPPER-107) and the functional spec:

> The format of the response body is a JSON object with the following fields:
> 
> * Output (JSON key name “output”): The prediction JSON object.
> * Default used (JSON key name “default”): A boolean set to true if the model did not return in time and the default output was used. False otherwise.
>
> If there is a non-recoverable error during query execution, an error JSON object will be returned instead and the HTTP response code will be set to the most appropriate error code (a 400 or 500 error code). The format of the error is a JSON object with the following fields:
> * Error (JSON key name “error”): The name of the error.
> * Cause (JSON key name “cause”): A string describing the source of the error.

Note: This PR is currently based on the branch corresponding to #89. As a result, many changes from #89 are redundantly included here.

### CODE REVIEW

File: src/frontends/src/query_frontend.hpp
Comment (dcrankshaw) [inline src/frontends/src/query_frontend.hpp:42]: Why are these `char*` instead of `std::string`? (I'm guessing there's a good reason)

File: src/libclipper/include/clipper/query_processor.hpp
Comment (dcrankshaw) [inline src/libclipper/include/clipper/query_processor.hpp]: Add `#include <stdexcept>` (it's better to explicitly include all the headers needed)

File: src/libclipper/include/clipper/query_processor.hpp
Comment (dcrankshaw) [inline src/libclipper/include/clipper/query_processor.hpp]: Let's define this in a separate `exceptions.hpp` file.

File: src/libclipper/src/query_processor.cpp
Comment (dcrankshaw) [inline src/libclipper/src/query_processor.cpp]: Move to `exceptions.cpp`

File: src/management/src/management_frontend.hpp
Comment (dcrankshaw) [inline src/management/src/management_frontend.hpp]: s/number/float

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
This looks great. We should wait for #89 to be merged first.

Does this require any changes to `clipper_manager.py` or the tutorial (beyond those needed for #89)?

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [reply src/frontends/src/query_frontend.hpp:42]: Our utility methods for adding key-value pairs to JSON objects expect `char*` keys. There is no implicit cast from `std::string` to `char*`, so I've defined the constants this way. Alternatively, I could define these with the `std::string` type and call `.data()` to get char* pointers, but this seems less readable.

File: src/management/src/management_frontend.hpp
Comment (Corey-Zumar) [reply src/management/src/management_frontend.hpp]: Done

**Review by Corey-Zumar — COMMENTED** [review]
@dcrankshaw Addressed your comments - moved `PredictError` to separate `exceptions.hpp` and `exceptions.cpp` files.

File: src/libclipper/include/clipper/query_processor.hpp
Comment (Corey-Zumar) [reply src/libclipper/include/clipper/query_processor.hpp]: Done

File: src/libclipper/include/clipper/query_processor.hpp
Comment (Corey-Zumar) [reply src/libclipper/include/clipper/query_processor.hpp]: Included `stdexept` in `exceptions.hpp`

File: src/libclipper/src/query_processor.cpp
Comment (Corey-Zumar) [reply src/libclipper/src/query_processor.cpp]: Done

**Comment by dcrankshaw** [comment]
LGTM. We need to wait on #115 and then #89 before merging this though

**Review by dcrankshaw — APPROVED** [review]

File: src/libclipper/src/query_processor.cpp
Comment (dcrankshaw) [inline src/libclipper/src/query_processor.cpp]: @Corey-Zumar Can you add a ratio counter here that tracks the fraction of predictions that use the default output?

**Comment by Corey-Zumar** [comment]
@dcrankshaw Rebased on `develop`. Please take another look over the code and confirm that the tutorial runs properly before merging - it's working well for me.

File: src/libclipper/src/query_processor.cpp
Comment (Corey-Zumar) [reply src/libclipper/src/query_processor.cpp]: Done

**Comment by dcrankshaw** [comment]
LGTM. Confirming that the tutorial works for me as well.

**Pull request merged**
